sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/CertificateValidationCheck.CSharp9.cs

```diff
@@ -0,0 +1,149 @@
+            }
+        }
+
+        #endregion inner assignment
+
+        #region Constructor Arguments
+
+        void ConstructorArguments()
+        {
+            OptionalConstructorArguments optA = new(this, cb: InvalidValidation);   // FN [flow-invalid-1]
+            OptionalConstructorArguments optB = new(this, cb: CompliantValidation);
+
+            using (var ms = new System.IO.MemoryStream())
+            using (System.Net.Security.SslStream ssl = new (ms, true, (sender, chain, certificate, SslPolicyErrors) => true)) // FN
+            {
+            }
+        }
+
+        record OptionalConstructorArguments
+        {
+            public OptionalConstructorArguments(object owner, int a = 0, int b = 0, RemoteCertificateValidationCallback cb = null)
+            {
+
+            }
+        }
+
+        #endregion Constructor Arguments
+
+        void DelegateReturnedByFunction()
+        {
+            CreateRQ().ServerCertificateValidationCallback += FindInvalid(false);       //Noncompliant [flow-FindInvalid]
+            CreateRQ().ServerCertificateValidationCallback += FindLambdaValidator();    //Noncompliant [flow-lambda]
+        }
+
+        static RemoteCertificateValidationCallback FindLambdaValidator()
+        {
+            return (_, _, _, _) => true;        //Secondary [flow-lambda]
+        }
+
+        static RemoteCertificateValidationCallback FindInvalid(bool useDelegate)   //All paths return noncompliant
+        {
+            if (useDelegate)
+            {
+                return InvalidValidation;                                  //Secondary [flow-FindInvalid]
+            }
+            else
+            {
+                return (sender, certificate, chain, SslPolicyErrors) => true;   //Secondary [flow-FindInvalid]
+            }
+        }
+
+        void GenericHandlerSignature()
+        {
+            //Generic signature check without RemoteCertificateValidationCallback
+            var ShouldTrigger = new RelatedSignatureType();
+            ShouldTrigger.Callback += InvalidValidation;                                           //Noncompliant [flow-invalid-2]
+            ShouldTrigger.Callback += CompliantValidation;
+
+            var ShouldNotTrigger = new NonrelatedSignatureType();
+            ShouldNotTrigger.Callback += (sender, chain, certificate, SslPolicyErrors) => true;   //Compliant, because signature types are not in expected order for validation
+            ShouldNotTrigger.Callback += (sender, chain, certificate, SslPolicyErrors) => false;
+        }
+
+        void DelegateReturnedByProperty()
+        {
+            CreateRQ().ServerCertificateValidationCallback += DelegateProperty;
+        }
+
+        static RemoteCertificateValidationCallback DelegateProperty
+        {
+            get
+            {
+                return (_, _, _, _) => true;   //False negative
+            }
+        }
+
+        #region helpers
+
+        static HttpWebRequest CreateRQ()
```

Comment:
now that I am looking at this, I could extract in a common file and pass that file to both pre-C#9 and C#9 tests